### PR TITLE
vg: add path primitives QuadTo and CurveTo

### LINKED ISSUE #482: vg: missing quadratic and cubic path elements
Is there a way to use quadratic and cubic bezier elements in vg?
From looking at the source, it seems that vg.Path does not provide these primitives.
As they are elementary to 2d vector graphics, I think they should be present.

I propose to add the constant:
```go
const (
	MoveComp = iota
	LineComp
	ArcComp
	CurveComp // This is new
	CloseComp
)
```

And the methods
```go
// QuadTo adds a quadratic curve element to the path defined by the control point p1 and the endpoint pt.
func (p *Path) QuadTo(p1, pt Point){...}

// CubeTo adds a cubic curve element to the path defined by the control points p1 and p2 and the endpoint pt.
func (p *Path) CubeTo(p1, p2, pt Point){...}
```

All backends should be able to handle the primitives:

- eps: curveto which is like CubeTo. QuadTo can be implemented by setting p1=p2
  - https://www.math.ubc.ca/~cass/graphics/manual/pdf/ch6.pdf (page 2, sec 6.2)
- img: draw2d has QuadCurveTo and CubicCurveTo
- pdf: gopdf has CurveTo and CurveBezierCubicTo
- svg: svgo has Qbez and Bezier
- tex: pgf has control points, which I hope are compatible
  - http://www.math.uni-leipzig.de/~hellmund/LaTeX/pgf-tut.pdf (page 8)

Did I miss anything?

What do you think?

### PR DESCRIPTION
**Pull request by ktye**
As discussed in #482, this adds bezier path elements to vg and all backends.
An example is provided in the [comment](https://github.com/gonum/plot/issues/482#issuecomment-437421668).

I tried to keep the style of the surrounding blocks.

closes #482

### CODE REVIEW

**Comment by ktye** [comment]
There have been 3 issues with CI:
- Code not gofmt simplified in: plotter/line.go
  - This should not be caused by this PR
- Recorder test:
  - has been updated.
  - Is the signature change an issue for serialization?
- Contour tests are failing.
  - can someone help out? It's not obvious for me, what to change.

**Comment by kortschak** [comment]
The gofmt failure is not a failure, but rather just a warning. We can fix that later.

The recorder type is primarily for testing. Serialisation should be robust though.

The contour test is failing because PathComp now contains a slice which cannot be compared by `==`. You will need to replace those comparisons with `reflect.DeepEqual`.

File: plotter/contour_test.go [hunk 1]
```diff
@@ -200,29 +200,29 @@ func circularPermutations(a, b vg.Path) bool {
 
 	var forward bool
 	for i, pc := range b {
-		if a[0] == pc {
+		if reflect.DeepEqual(a[0], pc) == true {
```
Comment (kortschak) [inline plotter/contour_test.go]: The `== true` is not necessary.

File: plotter/contour_test.go [hunk 2]
```diff
@@ -200,29 +200,29 @@ func circularPermutations(a, b vg.Path) bool {
...
 			off = i
 			forward = true
 			break
 		}
 	}
 	for i, pc := range a {
-		if b[(off+i)%len(a)] != pc {
+		if reflect.DeepEqual(b[(off+i)%len(a)], pc) == false {
```
Comment (kortschak) [inline plotter/contour_test.go]: Same here (use `!reflect.Deep...`).

File: plotter/contour_test.go [hunk 3]
```diff
@@ -200,29 +200,29 @@ func circularPermutations(a, b vg.Path) bool {
...
 			forward = false
 			break
 		}
 	}
 
 	var reverse bool
 	for i, pc := range b {
-		if a[0] == pc {
+		if reflect.DeepEqual(a[0], pc) == true {
```
Comment (kortschak) [inline plotter/contour_test.go]: And here.

File: plotter/contour_test.go [hunk 4]
```diff
@@ -200,29 +200,29 @@ func circularPermutations(a, b vg.Path) bool {
...
 			off = i
 			reverse = true
 			break
 		}
 	}
 	for i, pc := range a {
-		if b[(off-i+len(a))%len(a)] != pc {
+		if reflect.DeepEqual(b[(off-i+len(a))%len(a)], pc) == false {
```
Comment (kortschak) [inline plotter/contour_test.go]: And here.

File: vg/vg.go [hunk 5]
```diff
@@ -128,6 +128,18 @@ func (p *Path) Arc(pt Point, rad Length, s, a float64) {
 	})
 }
 
+// QuadTo adds a quadrative curve element to the path,
```
Comment (kortschak) [inline vg/vg.go]: s/quadrative/quadratic/

Comment (ktye) [reply vg/vg.go]: Done.
Thank you for review.

File: vg/vgimg/vgimg.go [hunk 6]
```diff
@@ -263,6 +263,17 @@ func (c *Canvas) outline(p vg.Path) {
 				comp.Radius.Dots(c.DPI()), comp.Radius.Dots(c.DPI()),
 				comp.Start, comp.Angle)
 
+		case vg.CurveComp:
+			if len(comp.Control) == 1 {
```
Comment (sbinet) [inline vg/vgimg/vgimg.go]: I would perhaps rewrite it as:
```go
switch len(comp.Control) {
case 1:
    // ...
case 2:
    // ...
}
```

File: vg/vgsvg/vgsvg.go [hunk 7]
```diff
@@ -171,6 +171,19 @@ func (c *Canvas) pathData(path vg.Path) string {
 			} else {
 				x, y = arc(buf, c, &comp)
 			}
+		case vg.CurveComp:
+			if len(comp.Control) == 1 {
```
Comment (sbinet) [inline vg/vgsvg/vgsvg.go]: ditto.

File: vg/vgpdf/vgpdf.go [hunk 8]
```diff
@@ -249,6 +249,17 @@ func (c *Canvas) pdfPath(path vg.Path, style string) {
 			c.doc.LineTo(c.pdfPoint(comp.Pos))
 		case vg.ArcComp:
 			c.arc(comp, style)
+		case vg.CurveComp:
+			if len(comp.Control) > 0 {
+				cx, cy := c.pdfPoint(comp.Control[0])
+				px, py := c.pdfPoint(comp.Pos)
+				if len(comp.Control) == 1 {
```
Comment (sbinet) [inline vg/vgpdf/vgpdf.go]: ditto.

Comment (ktye) [reply vg/vgpdf/vgpdf.go]: I interpret "I would perhaps" as to wait for another opinion. Otherwise please tell me, and I do it.

Comment (sbinet) [reply vg/vgpdf/vgpdf.go]: yep. let's wait for @kortschak to chime in.

Comment (kortschak) [reply vg/vgpdf/vgpdf.go]: Yes, this improves readability.

Comment (ktye) [reply vg/vgpdf/vgpdf.go]: Ok, it's done.

File: vg/vgtex/canvas.go [hunk 9]
```diff
@@ -258,6 +258,16 @@ func (c *Canvas) wpath(p vg.Path) {
 			angle := comp.Angle * degPerRadian
 			r := comp.Radius
 			c.wtex(`\pgfpatharc{%g}{%g}{%gpt}`, start, angle, r)
+		case vg.CurveComp:
+			if len(comp.Control) > 0 {
```
[the remark this answers is not in the document]
Comment (ktye) [reply vg/vgtex/canvas.go]: Would somthing like that be ok?
```go
panic(fmt.Sprintf("CurveComp cannot have %d number of control points", len(comp.Control)))
```

[the remark this answers is not in the document]
Comment (kortschak) [reply vg/vgtex/canvas.go]: `panic("<pkg>: invalid number of control points")` should be fine where `<pkg>` is set for each package.

[the remark this answers is not in the document]
Comment (kortschak) [reply vg/vgtex/canvas.go]: You missed this one.

**Comment by kortschak** [comment]
This is almost ready. Would you please also send a PR to gonum/gonum adding yourself to AUTHORS and CONTRIBUTORS with the commit message "A+C: add <your name>"?

File: vg/vgtex/canvas.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kortschak) [reply vg/vgtex/canvas.go]: Thanks

**Comment by kortschak** [comment]
Thanks

**Pull request merged**